pkg/resource/deploy/source_eval.go

```diff
@@ -1567,6 +1619,22 @@ func (rm *resmon) RegisterStackInvokeTransform(ctx context.Context, cb *pulumirp
 	return &emptypb.Empty{}, nil
 }
 
+// WaitForShutdown blocks until the resource monitor is canceled, which will
+// happen once all the steps have executed. This allows the language runtime to
+// stay running and handle callback requests even after the user program has
+// completed. This can only be called once.
+func (rm *resmon) WaitForShutdown(ctx context.Context, req *emptypb.Empty) (*emptypb.Empty, error) {
+	logging.V(6).Infof("WaitForShutdown waiting ...")
+	if rm.hasWaiter.CompareAndSwap(false, true) {
+		rm.finChan <- nil        // Let the source iterator know there will be no more events ...
```

Comment:
I think we know this at program shutdown time, not now

Reply:
If we don't signal this here, we won't return from the loop that calls `src.Next()` in `deploymentExecutor.Execute`, and so we never return from that function. This means we never call the `defer src.Close()`. Without that we won't call `resmon.Cancel()`, and so we never write to `resmon.waitForShutdownChan`. The result of that is that `WaitForShutdown` never returns and the program never exits. We're deadlocked.